### PR TITLE
Refactor jtx::Env

### PR DESCRIPTION
**Pull request by vinniefalco**
These changes eliminate the Env's OpenLedger member and make transactions go through the Application associated with each instance of the Env, making the unit tests follow a code path closer to the production code path.

Reviewers: @miguelportilla @seelabs

### CODE REVIEW

File: src/ripple/test/jtx/Env.h [hunk 1]
```diff
@@ -377,7 +456,7 @@ class Env
 
     /** Return the TER for the last JTx. */
     TER
-    ter() const
+    ter()
```
Comment (seelabs) [inline src/ripple/test/jtx/Env.h]: I understand that C++-11 has a stricter definition of `const` than previous versions. However, most read-only functions should still be marked as `const`. Many of the functions here should still be `const`. By adding an `Application::openLedger() const` function we can bring back `const`. Here's a branch that shows what I have in mind: https://github.com/ripple/rippled/commit/52e92331c155857dedb2cb0c92fd46ae9388b6dc

File: src/ripple/test/jtx/Env.h [hunk 2]
```diff
@@ -60,38 +69,37 @@ namespace detail {
...
-    std::array<Account, sizeof...(Args)>
+    template <class U, class ...Args>
+    std::array<T, sizeof...(Args)>
     flatten (Args&& ...args)
     {
-        return std::array<T,
-            sizeof...(Args)> {
-                std::forward<Args>(args)...};
+        return std::array<U,sizeof...(Args)> {
+            std::forward<Args>(args)...};
```
Comment (seelabs) [inline src/ripple/test/jtx/Env.h]: I'm not sure what's going on in this workaround. `U` should always be the same as `T` and `sizeof...(Args)` should always be the same as `N`. I'd like to get rid of this complexity and replace `vararg_array_helper` with just `std:array`. We can use a type alias or just replace the current uses with `std::array`. Here's a branch that shows what I have in mind: https://github.com/ripple/rippled/commit/6281e3be09a310fc2625a7f0bddc5d1e69990698

File: src/ripple/test/jtx/impl/Account.cpp [hunk 3]
```diff
@@ -36,15 +44,28 @@ Account::Account(std::string name,
...
+        human_ = toBase58(id_);
+        cache_[p] = *this;
+    }
+    else
+    {
+        pk_ = iter->second.pk_;
+        sk_ = iter->second.sk_;
+        id_ = iter->second.id_;
+        human_ = iter->second.human_;
+    }
```
Comment (seelabs) [inline src/ripple/test/jtx/impl/Account.cpp]: We can simplify this code by using a delegating to the move constructor and a helper function that returns the account from the cache & add it to the cache if needed. Code: https://github.com/ripple/rippled/commit/4ea8c5617bef844e78ea6fdd8de61b6f8d461891

Note: That code required making one of the private constructors public, so I could efficiently emplace into a map. If we need to keep that constructor private I'll modify this solution.

File: src/ripple/test/jtx/Env.h [hunk 4]
```diff
@@ -103,13 +111,20 @@ struct noripple_helper
...
+/** Activate features in the Env ctor */
+template <class... Args>
+detail::vararg_array_helper<uint256, 1 + sizeof...(Args)>
+features (uint256 const& key, Args const&... args)
 {
-    return detail::noripple_helper<
-        1 + sizeof...(Args)>(
-            account, args...);
+    return detail::vararg_array_helper<uint256,
+        1 + sizeof...(Args)>(key, args...);
```
Comment (seelabs) [inline src/ripple/test/jtx/Env.h]: `return {{key, args...}};` works too (tested with the `std::array` modification posted earlier)

File: src/ripple/test/jtx/Env.h [hunk 5]
```diff
@@ -135,27 +145,70 @@ class Env
...
+    void
+    construct()
+    {
+    }
+
+    template <class Arg, class... Args>
+    void
+    construct (Arg&& arg, Args&&... args)
+    {
+        construct_arg(arg);
```
Comment (seelabs) [inline src/ripple/test/jtx/Env.h]: `std::forward<Args>(arg)`

File: src/ripple/basics/impl/make_SSLContext.cpp [hunk 6]
```diff
@@ -514,15 +514,20 @@ initAuthenticated (boost::asio::ssl::context& context,
...
-        std::make_shared<boost::asio::ssl::context> (
-            boost::asio::ssl::context::sslv23);
-    // By default, allow anonymous DH.
-    openssl::detail::initAnonymous (
-        *context, "ALL:!LOW:!EXP:!MD5:@STRENGTH");
-    // VFALCO NOTE, It seems the WebSocket context never has
-    // set_verify_mode called, for either setting of WEBSOCKET_SECURE
-    context->set_verify_mode (boost::asio::ssl::verify_none);
+    static auto const context =
+        []()
```
Comment (seelabs) [inline src/ripple/basics/impl/make_SSLContext.cpp]: Nit: I prefer `[]` over `[]()` for lambdas with no parameters and an inferred return type.

File: src/ripple/test/jtx/Account.h [hunk 7]
```diff
@@ -115,6 +109,20 @@ class Account
...
     IOU
     operator[](std::string const& s) const;
+
+private:
+    Account (std::string name,
+        std::pair<PublicKey, SecretKey> const& keys);
+
+    static std::unordered_map<
+        std::pair<std::string, KeyType>,
+            Account, beast::uhash<>> cache_;
```
Comment (seelabs) [inline src/ripple/test/jtx/Account.h]: `cashe_` could be a static inside the constructor (or inside the `fromCache` helper function if you take that change).

File: src/beast/beast/threads/Stoppable.h [hunk 8]
```diff
@@ -314,22 +324,56 @@ class RootStoppable : public Stoppable
...
+template <class Rep, class Period>
+bool
+RootStoppable::alertable_sleep_for(
+    std::chrono::duration<Rep, Period> const& d)
+{
+    std::unique_lock<std::mutex> lock(m_);
+    if (m_calledStop)
+        return true;
+    return c_.wait_for(lock, d) ==
+        std::cv_status::no_timeout;
```
Comment (seelabs) [inline src/beast/beast/threads/Stoppable.h]: Without a predicate, `wait_for` can have spurious wakeups. Given this is a sleep function, and looking at the callers, this looks OK to me. However, a spurious wakeup would make the return value incorrect. I'd suggest two changes:

1) Change the return type to `void`
2) Add a comment `// spurious wakeups OK` by the `wait_for` function.

Comment (vinniefalco) [inline src/beast/beast/threads/Stoppable.h]: Good call! This is a bug. It should have a predicate.

File: src/ripple/test/jtx/Env.h (hunk 2 above)
Comment (seelabs) [inline src/ripple/test/jtx/Env.h]: @vinniefalco @nbougalis Confirmed that the `std::array` change proposed above works on windows.

**Comment by seelabs** [comment]
@vinniefalco Here are a set of commits for you to cherry pick. These have been tested on both linux and windows:

1) Mark read-only functions as `const`: https://github.com/ripple/rippled/commit/52e92331c155857dedb2cb0c92fd46ae9388b6dc
2) Replace vararg_array_helper with std::array: https://github.com/ripple/rippled/commit/0eeae7b99a379c7d839b8098e64a36bf5d11620d
3)Use delegating ctor to add account to cache: https://github.com/ripple/rippled/commit/ed568f0c6e5f3353b6c251d80a76182fc3f4e29d
4) Add predicate to `wait_for` to protect against spurious wakeups: https://github.com/ripple/rippled/commit/8dbec58a29bbe82ab384a35051a91f4297a08054
5) m_calledStop does not need to be atomic: https://github.com/ripple/rippled/commit/ebb583d1c4d05253204edc5e2e6e13030aa01ee1

I know you disagree with the `const` commit. Commit (5) is a new issue I found while adding the predicate to the `wait_for`. I'm also happy to squash any subset of these into a single commit if it makes your life easier.

File: src/ripple/app/tests/Offer.test.cpp [hunk 9]
```diff
@@ -149,8 +149,11 @@ class Offer_test : public beast::unit_test::suite
...
-        for (auto timeDelta : {-10s, 10s}){
-            NetClock::time_point const closeTime = switchoverTime + timeDelta;
+        for (auto timeDelta : {
+            - env.closed()->info().closeTimeResolution,
+                env.closed()->info().closeTimeResolution} )
+        {
+            auto const closeTime = switchoverTime + timeDelta;
             STAmountCalcSwitchovers switchover (closeTime);
             env.close (closeTime);
             // Will fail without the underflow fix
```
Comment (ximinez) [inline src/ripple/app/tests/Offer.test.cpp]: Something is causing the test case on line 162 to fail in Windows.
```
            env (pay ("alice", "bob", EUR (epsilon)), path (~EUR),
                sendmax (USD (100)), ter (expectedResult));
```

Comment (ximinez) [inline src/ripple/app/tests/Offer.test.cpp]: The cause is _not_ the immediate change above. I reverted those couple of lines and it still fails.

Comment (ximinez) [inline src/ripple/app/tests/Offer.test.cpp]: It's the first pass that fails: `expectedResult == tecPATH_PARTIAL`, but the result is `tesSUCCESS`.

Comment (ximinez) [inline src/ripple/app/tests/Offer.test.cpp]: This failure was caused by a time flag rolling over on the morning of 12/28. It is resolved in https://github.com/ximinez/rippled/commit/ce0290351498a92d2941ec3aa937e31e6c07c12b, which needs to be merged / cherry picked into this branch.

**Pull request closed without merging**